### ISSUE TITLE
For clustering indexes, run query filters within bulk fetch callback and avoid copying unnecessary documents into the RowBuffer

### ISSUE THREAD

**Issue by zkasheff** [Tokutek/mongo#751]
#725 showed us a valuable way to improve the performance of range queries: if we can avoid an unnecessary copy of the row from the fractal tree to the RowBuffer, we can improve performance. There are many ways to apply this, but this ticket focuses on a single one: query filtering.

For clustering (and possibly non-clustering) indexes, if we can learn that the cursor's caller (query, aggregation framework, whatever) will filter the document because it does not match the query, then we should avoid copying it into the RowBuffer in the first place. For example, if we have a clustering index on a, and want to find documents where a is 10 and b is 100, then as we search the index where a is 10, within the bulk fetch callback, we should be smart enough to filter documents where b is not 100 and only copy the documents into the RowBuffer where b is 100.

In other words, this is index condition pushdown done right.
